plaidml/edsl/edsl_doc.cc

```diff
@@ -0,0 +1,338 @@
+Tensor MaxPool1D(const Tensor& I) {
+  TensorDim N;
+  TensorIndex i, j;
+  I.bind_dims(N);
+  auto O = TensorOutput(N / 2);
+  O(i) >= I(2 * i + j);
+  O.add_constraint(j < 2);
+  return O;
+}
+// max_pool_1d_end
+
+// max_pool_1d_odd_start
+Tensor MaxPool1DOdd(const Tensor& I) {
+  TensorDim N;
+  TensorIndex i, j;
+  I.bind_dims(N);
+  auto O = TensorOutput((N + 1) / 2);
+  O(i) >= I(2 * i + j);
+  O.add_constraint(j < 2);
+  return O;
+}
+// max_pool_1d_odd_end
+
+// skip_start
+Tensor Skip(const Tensor& I) {
+  TensorDim M, N;
+  TensorIndex i, j;
+  I.bind_dims(M, N);
+  auto O = TensorOutput(N);
+  O(2 * i) += I(2 * i, j);
+  return O;
+}
+// skip_end
+
+// cumsum_start
+Tensor CumSum(const Tensor& I) {
+  TensorDim N;
+  TensorIndex i, k;
+  I.bind_dims(N);
+  auto O = TensorOutput(N);
+  O(i) += I(k);
+  O.add_constraint(i - k < N);
+  return O;
+}
+// cumsum_end
+
+// conv_1d_start
+Tensor Conv1D(const Tensor& I, const Tensor& K) {
+  TensorDim N, X, KX, CI, CO;
+  TensorIndex n, x, k, ci, co;
+  I.bind_dims(N, X, CI);
+  K.bind_dims(KX, CI, CO);
+  auto O = TensorOutput(N, X - KX + 1, CO);
+  O(n, x, co) += I(n, x + k, ci) * K(k, ci, co);
+  return O;
+}
+// conv_1d_end
+
+// conv_2d_dilated_start
+Tensor Conv2DDilated(const Tensor& I, const Tensor& K) {
+  TensorDim N, X, Y, KX, KY, CI, CO;
+  TensorIndex n, x, y, kx, ky, ci, co;
+  I.bind_dims(N, X, Y, CI);
+  K.bind_dims(KX, KY, CI, CO);
+  auto O = TensorOutput(N, X - 2 * (KX - 1), Y - 3 * (KY - 1), CO);
+  O(n, x, y, co) += I(n, x + 2 * kx, y + 3 * ky, ci) * K(kx, ky, ci, co);
+  return O;
+}
+// conv_2d_dilated_end
+
+// complex_conv_start
+Tensor ComplexConv2D(const Tensor& I, const Tensor& K,
+                     const std::vector<size_t>& s,  // stride coeffs
+                     const std::vector<size_t>& d   // dilation coeffs
+) {
+  // "same-lower" autopadding will be applied
+  TensorDim N, G, GCI, GCO;
+  std::vector<TensorDim> X(2);
+  std::vector<TensorDim> KD(2);
+  TensorIndex n, g, gci, gco;
+  std::vector<TensorIndex> x(2);
+  std::vector<TensorIndex> k(2);
+  I.bind_dims(N, X[0], X[1], G, GCI);
+  K.bind_dims(KD[0], KD[1], G, GCI, GCO);
+  // Compute output spatial dimensions
+  std::vector<TensorDim> Y(2);
+  for (size_t i = 0; i < Y.size(); ++i) {
+    Y[i] = (X[i] + s[i] - 1) / s[i];
+  }
+  // Compute the effective kernel size after dilation
+  std::vector<TensorDim> EK(2);
+  for (size_t i = 0; i < EK.size(); ++i) {
+    EK[i] = d[i] * (KD[i] - 1) + 1;
+  }
+  // Compute the padding offset
+  std::vector<TensorDim> P(2);
+  for (size_t i = 0; i < P.size(); ++i) {
+    P[i] = ((Y[i] - 1) * s[i] + EK[i] - X[i]) / 2;
+  }
+  // Specify the output size
+  auto O = TensorOutput(N, Y[0], Y[1], G, GCO);
+  // Compute the convolution
+  O(n, x[0], x[1], g, gco) +=
+      I(n, s[0] * x[0] + d[0] * k[0] - P[0], s[1] * x[1] + d[1] * k[1] - P[1], g, gci) * K(k[0], k[1], g, gci, gco);
+  return O;
+}
+// complex_conv_end
+
+TEST(CppEdsl, SumOveAxis) {
```

Comment:
Put these tests in some other fixture, we already have `CppEdsl` in another file.